http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/TlsHandshakeEnsuringChannelPool.java

```diff
@@ -0,0 +1,122 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.pool.ChannelPool;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.Promise;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.nio.netty.internal.utils.NettyClientLogger;
+
+/**
+ * This {@code ChannelPool} ensures that the channel has completed the TLS negotiation before giving it back to the caller. Note
+ * that it's possible for a channel to have multiple {@link SslHandler} instances its pipeline, for example if it's using a
+ * proxy over HTTPS. This pool explicitly looks for the handler of the given name.
+ * <p>
+ * If TLS setup fails, the channel will be closed and released back to the wrapped pool and the future will be failed.
+ */
+@SdkInternalApi
+public class TlsHandshakeEnsuringChannelPool implements ChannelPool {
+    private static final NettyClientLogger LOGGER = NettyClientLogger.getLogger(TlsHandshakeEnsuringChannelPool.class);
+    private final EventLoopGroup eventLoopGroup;
+    private final String sslHandlerName;
+    private final ChannelPool delegate;
+
+    public TlsHandshakeEnsuringChannelPool(EventLoopGroup eventLoopGroup, String sslHandlerName, ChannelPool delegate) {
+        this.eventLoopGroup = eventLoopGroup;
+        this.sslHandlerName = sslHandlerName;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Future<Channel> acquire() {
+        return acquire(eventLoopGroup.next().newPromise());
+    }
+
+    @Override
+    public Future<Channel> acquire(Promise<Channel> promise) {
+        Promise<Channel> delegatePromise = eventLoopGroup.next().newPromise();
+
+        delegate.acquire(delegatePromise).addListener((GenericFutureListener<Future<Channel>>)
+                                                       f -> tlsHandshakeListener(f, promise));
+
+        return promise;
+    }
+
+    @Override
+    public Future<Void> release(Channel channel) {
+        return delegate.release(channel);
+    }
+
+    @Override
+    public Future<Void> release(Channel channel, Promise<Void> promise) {
+        return delegate.release(channel, promise);
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    private void tlsHandshakeListener(Future<Channel> f, Promise<Channel> promise) throws ExecutionException,
+                                                                                                 InterruptedException {
+        if (!f.isSuccess()) {
+            promise.tryFailure(f.cause());
+            return;
+        }
```

Comment:
Should we more explicitly propagate cancellation here?  E.g., via `NettyUtils#consumeOrPropagate`.